include/gdsync/core.h

```diff
@@ -39,37 +39,86 @@
     ( ((((v) & 0xffff0000U) >> 16) == GDS_API_MAJOR_VERSION) &&   \
       ((((v) & 0x0000ffffU) >> 0 ) >= GDS_API_MINOR_VERSION) )
 
+#include <infiniband/mlx5dv.h>
+
 typedef enum gds_param {
-    GDS_PARAM_VERSION,
-    GDS_NUM_PARAMS
+        GDS_PARAM_VERSION,
+        GDS_NUM_PARAMS
 } gds_param_t;
 
 int gds_query_param(gds_param_t param, int *value);
 
 enum gds_create_qp_flags {
-    GDS_CREATE_QP_DEFAULT      = 0,
-    GDS_CREATE_QP_WQ_ON_GPU    = 1<<0,
-    GDS_CREATE_QP_TX_CQ_ON_GPU = 1<<1,
-    GDS_CREATE_QP_RX_CQ_ON_GPU = 1<<2,
-    GDS_CREATE_QP_WQ_DBREC_ON_GPU = 1<<5,
+        GDS_CREATE_QP_DEFAULT      = 0,
+        GDS_CREATE_QP_WQ_ON_GPU    = 1<<0,
+        GDS_CREATE_QP_TX_CQ_ON_GPU = 1<<1,
+        GDS_CREATE_QP_RX_CQ_ON_GPU = 1<<2,
+        GDS_CREATE_QP_WQ_DBREC_ON_GPU = 1<<5,
+};
+
+typedef struct ibv_qp_init_attr gds_qp_init_attr_t;
+typedef struct ibv_send_wr gds_send_wr;
+
+struct gds_peek_entry {
+        uint32_t busy;
+        uint32_t next;
 };
 
-typedef struct ibv_exp_qp_init_attr gds_qp_init_attr_t;
-typedef struct ibv_exp_send_wr gds_send_wr;
+typedef enum gds_cq_type {
+        GDS_CQ_TYPE_SQ,
+        GDS_CQ_TYPE_RQ
+} gds_cq_type_t;
 
 struct gds_cq {
-        struct ibv_cq *cq;
-        uint32_t curr_offset;
+        struct ibv_cq          *cq;
+        uint32_t                curr_offset;
```

Comment:
do we really need to expose the internal implementation? 
for example, by using struct mlx5dv_cq, you introduce a compile time dependency on the user code

Reply:
can you wrap the stuff coming from peer-direct async implementation into a mlx5 specific cq struct ?